Basic Redstone Components

xNatsuri opened the pull request:
This pull request aims to have basic redstone components. this pr was built from #739.

- levers
- redstone wire
- redstone block
- redstone torch

- RedstoneUpdate world Handler

xNatsuri committed:
implement basic redstone components

xNatsuri committed:
Merge branch 'df-mc:master' into feature/basic-redstone

xNatsuri committed:
remove unused functions

xNatsuri committed:
Merge branch 'df-mc:master' into feature/basic-redstone

xNatsuri committed:
Merge branch 'df-mc:master' into feature/basic-redstone

TwistedAsylumMC committed:
Merge branch 'master' into feature/basic-redstone

xNatsuri committed:
remove redstone torches will be added in future pr.

xNatsuri committed:
add back redstone torches

xNatsuri committed:
feat: implement torch burnout

xNatsuri committed:
match to java

xNatsuri committed:
match to java

xNatsuri committed:
match to java

xNatsuri committed:
was matching bedrock whoops

xNatsuri committed:
fix redstone block logic

xNatsuri committed:
Merge branch 'master' into feature/basic-redstone

xNatsuri committed:
requested changes

xNatsuri committed:
remove unused interface

xNatsuri committed:
requested changes

DaPigGuy commented on server/block/redstone_wire.go:
```diff
@@ -0,0 +1,189 @@
+	if !used {
+		return
+	}
+	if _, ok := tx.Liquid(pos); ok {
+		return false
+	}
```
```suggestion
```
This is handled elsewhere now

xNatsuri committed:
requested changes

xNatsuri committed:
requested changes

xNatsuri committed:
door interaction with redstone

xNatsuri committed:
note block interaction with redstone

xNatsuri committed:
requested changes

xNatsuri committed:
revert changes

xNatsuri committed:
fix updateAroundRedstone func

xNatsuri committed:
Merge branch 'feature/basic-redstone' of https://github.com/xNatsuri/dragonfly into feature/basic-redstone

DaPigGuy commented on server/block/note.go:
Note blocks have a powered state. It is probably stored in the block entity NBT but does not appear to be documented in the wiki

DaPigGuy commented on server/block/note.go:
```diff
@@ -48,13 +49,40 @@ func (n Note) EncodeNBT() map[string]any {
+}
+
+// powered checks if the note block is receiving redstone power.
+func (n Note) powered(pos cube.Pos, tx *world.Tx) bool {
+	for _, face := range cube.Faces() {
+		adjacentPos := pos.Side(face)
+		if power := tx.RedstonePower(adjacentPos, face.Opposite(), true); power > 0 {
+			return true
+		}
+		if power := tx.RedstonePower(adjacentPos, face.Opposite(), false); power > 0 {
+			return true
+		}
+	}
+	return false
 }
```
Might be a good idea to have a generic version of this? Lots of blocks, e.g. redstone lamps, TNT, dispensers need this exact same logic

Doors can also use this generic method, e.g. `powered(bottomPos, tx) || powered(topPos, tx)`

DaPigGuy replied:
Also why are we checking redstone power without account for dust here?

xNatsuri committed:
requested changes

xNatsuri committed:
add RedstoneUpdate world handler

xNatsuri committed:
Merge branch 'master' into feature/basic-redstone

xNatsuri committed:
Merge branch 'master' into feature/basic-redstone

xNatsuri committed:
implement redstone trim and update docs

xNatsuri committed:
Merge branch 'master' into feature/basic-redstone

xNatsuri committed:
fix redstone torch logic

xNatsuri committed:
Merge branch 'master' into feature/basic-redstone

xNatsuri committed:
Merge branch 'master' into feature/basic-redstone

xNatsuri committed:
Merge branch 'master' into feature/basic-redstone

HashimTheArab committed:
Merge branch 'master' into feature/basic-redstone

HashimTheArab committed:
Post fixes for basic redstone implementation (#2)

* fix: incorrect redstone block strongpower

* simplify lever.go logic

* fix: Register the inventory form of redstone dust as item.RedstoneWire instead of block.RedstoneWire so decoded redstone stacks keep armour trim material behavior

* fix: fire redstone update events during updates instead of power reads. Also simplify code

We removed it from Tx.RedstonePower because RedstonePower is used as a read/query during redstone calculations.

  If the handler stays inside RedstonePower, then every time the algorithm asks “how much power is this neighbor emitting?”, it also fires a cancellable redstone update event.

  That is problematic because:

  - A single redstone update may call RedstonePower many times.
  - The handler fires for blocks that are merely being sampled, not actually updated.
  - Cancelling the event makes the query return 0, so reading power can change calculation results.
  - It makes behavior depend on internal calculation order.

  So I moved it to updateRedstone, where the code is actually about to call:

  r.RedstoneUpdate(pos, tx)

  That way:

  - Tx.RedstonePower = “read current power”
  - HandleRedstoneUpdate = “a redstone-updatable block is about to update”
  - cancellation skips the update, not arbitrary power sampling

  This matches the event name better and avoids side effects during power reads.

* fix: a redstone torch produces full-strength signal on all adjacent sides except for its attached block, and can power the block directly above it

* refactor(redstone torch): remove unused field and improve code readability

* fix(wooden door): wiki says that wooden doors can be powered by redstone dust, also simplify logic

* feat: add powered state to noteblocks, add sound packet handling, simplify code and document methods

Honor note block powered state during redstone updates

Track the note block powered state in block entity NBT so redstone updates only play the note when it first becomes powered. This matches Bedrock's block-entity-backed note block behavior and avoids retriggering while power stays on.

* Preserve formatting after upstream redstone merge

The source branch merge removed the final newline from server/world/tx.go, which caused gofmt/lint to fail. Restore gofmt output without changing behavior.\n\nConstraint: xNatsuri feature/basic-redstone was merged after local review fixes\nConfidence: high\nScope-risk: narrow\nTested: gofmt -w server/world/tx.go\nNot-tested: Full suite after formatting-only fix

* improve doc

HashimTheArab committed:
add redstone ore (#3)

* fix: incorrect redstone block strongpower

* simplify lever.go logic

* fix: Register the inventory form of redstone dust as item.RedstoneWire instead of block.RedstoneWire so decoded redstone stacks keep armour trim material behavior

* fix: fire redstone update events during updates instead of power reads. Also simplify code

We removed it from Tx.RedstonePower because RedstonePower is used as a read/query during redstone calculations.

  If the handler stays inside RedstonePower, then every time the algorithm asks “how much power is this neighbor emitting?”, it also fires a cancellable redstone update event.

  That is problematic because:

  - A single redstone update may call RedstonePower many times.
  - The handler fires for blocks that are merely being sampled, not actually updated.
  - Cancelling the event makes the query return 0, so reading power can change calculation results.
  - It makes behavior depend on internal calculation order.

  So I moved it to updateRedstone, where the code is actually about to call:

  r.RedstoneUpdate(pos, tx)

  That way:

  - Tx.RedstonePower = “read current power”
  - HandleRedstoneUpdate = “a redstone-updatable block is about to update”
  - cancellation skips the update, not arbitrary power sampling

  This matches the event name better and avoids side effects during power reads.

* fix: a redstone torch produces full-strength signal on all adjacent sides except for its attached block, and can power the block directly above it

* refactor(redstone torch): remove unused field and improve code readability

* fix(wooden door): wiki says that wooden doors can be powered by redstone dust, also simplify logic

* feat: add powered state to noteblocks, add sound packet handling, simplify code and document methods

Honor note block powered state during redstone updates

Track the note block powered state in block entity NBT so redstone updates only play the note when it first becomes powered. This matches Bedrock's block-entity-backed note block behavior and avoids retriggering while power stays on.

* Preserve formatting after upstream redstone merge

The source branch merge removed the final newline from server/world/tx.go, which caused gofmt/lint to fail. Restore gofmt output without changing behavior.\n\nConstraint: xNatsuri feature/basic-redstone was merged after local review fixes\nConfidence: high\nScope-risk: narrow\nTested: gofmt -w server/world/tx.go\nNot-tested: Full suite after formatting-only fix

* improve doc

* feat: redstone ore

* fix: make sure bottom face is solid for redstone wire

* Avoid stale redstone torch burnout stat

HashimTheArab committed:
Fix adjacent redstone sources powering dust

  Use the source-facing side when redstone dust queries neighbouring
  power so wall torches and levers emit toward adjacent dust correctly.

HashimTheArab committed:
fix: register block RedstoneWire instead of item and add TrimMaterial and MaterialColour methods to block

- item.RedstoneWire is kept for item-package trim registration; block.RedstoneWire is registered for placement. Since smithing checks the registered/decoded item, block.RedstoneWire also needs to satisfy ArmourTrimMaterial, delegating to item.RedstoneWire to avoid duplicated trim data.

HashimTheArab committed:
fix(redstone_ore.go): use block RedstoneWire type instead of item and shorten docs in redstone_wire.go

HashimTheArab committed:
fix(redstone_block.go): weak power is 15 and strong power is 0

HashimTheArab committed:
fix(redstone_ore): remove NeighbourUpdateTick

HashimTheArab committed:
fix: fix receiver-relative redstone power checks for adjacent components

Several adjacent redstone power checks were using the wrong face frame, which caused wall torches and similar sources to be missed by wires, doors, and note blocks. This change standardizes those lookups on the receiver-relative face convention, updates redstone torch weak-power behavior to match, and adds regression coverage for the affected cases.

HashimTheArab committed:
fix(wood door): only check horizontal faces in powered

HashimTheArab committed:
refactor: revert wooden doors to master branch

HashimTheArab committed:
fix: single dust with no horizontal connections should power all sides. Also fixes stepdown logic and power-through-block

example: dust is being powered by a vertical connection (such as a redstone block placed above it) and a trapdoor next to the dust

additional:
  Redstone dust and solid-block update propagation had a few stale-state
  edges: standalone powered dust did not provide horizontal weak power to
  passive receivers, mechanisms behind indirectly powered solid blocks did
  not always receive redstone updates, and wire step-down behavior was tied
  to light diffusion instead of redstone-specific travel rules.

  This adds a small step-down capability for blocks with special downward
  wire behavior while preserving the default full-face-solid fallback.

HashimTheArab committed:
fix: glowstone should be transparent, it does not transmit power. connecting a lever and a noteblock to glowstone on opposite sides should not power the note block, but it does with opaque blocks such as stone.

This also matches minecraft behavior because glowstone emits light and does not block light like opaque stone

HashimTheArab committed:
fix: handle redstone dust vertical travel through transparent supports

HashimTheArab committed:
fix: redstone torch is always a redstone source, whether it provides power is handled WeakPower and StrongPower

HashimTheArab committed:
fix: call SetBlock before internal blocks BreakHandler

fixes this case:
When dust breaks because its support disappears, breakBlock runs the dust BreakHandler before the block is actually set to air. That means updateStrongRedstone observes the old powered dust, then the later air placement only sends generic neighbour updates. Passive redstone receivers like Note only implement RedstoneUpdate, so they can stay powered after the dust is gone. Fix by propagating redstone again after breakBlock returns, or by making this break path honor the BreakHandler “after broken” contract.

HashimTheArab committed:
fix: redstone wire updates bypassed HandleRedstoneUpdate and redstone_wire NeighbourUpdateTick called r.RedstoneUpdate directly instead of updateRedstone to go through the handler

HashimTheArab committed:
refactor: move torch state management into internal/redstone (#4)

* refactor: move torch state management into internal/redstone

* fix: store by world current tick instead of time because Time() is effected by world.StopTime()

* fix: powered noteblocks should also power adjacent noteblocks

HashimTheArab committed:
fix: allow vertical power both ways

RestartFU commented on server/block/redstone.go:
```diff
@@ -0,0 +1,409 @@
+
+	if i != j {
+		wire.Power = j
+		tx.SetBlock(node.pos, wire, &world.SetOpts{DisableBlockUpdates: true})
```
do we not want to update the node's block state aswell?

HashimTheArab replied:
The node cache is updated by the caller. calculateCurrentChanges returns the updated wire, and updateNode assigns it back with node.block = newWire before propagating, so updating node.block inside this helper would be redundant.

RestartFU commented on server/block/redstone.go:
```diff
@@ -0,0 +1,409 @@
+		}
+	}
+
+	j := blockPower - 1
```
maybe add a check if blockPower is 0, set j to 0?

HashimTheArab replied:
added

RestartFU commented on server/block/redstone.go:
```diff
@@ -0,0 +1,409 @@
+			neighbourBlock := neighbour.block
+			_, neighbourSolid := neighbourBlock.Model().(model.Solid)
+
+			blockPower = n.maxCurrentStrength(neighbourBlock, blockPower)
```
add a check, break if blockPower >= 15?

HashimTheArab replied:
good catch

HashimTheArab committed:
refactor: add max 0 check

HashimTheArab committed:
add blockpower == 15 check

RestartFU commented on server/block/redstone_wire.go:
```diff
@@ -0,0 +1,237 @@
+// NeighbourUpdateTick ...
+func (r RedstoneWire) NeighbourUpdateTick(pos, neighbour cube.Pos, tx *world.Tx) {
+	if pos == neighbour {
+		// Ignore neighbour updates on ourself.
```
unless I'm wrong?

```suggestion
		// Ignore neighbour updates on ourselves.
```

HashimTheArab replied:
resolved

HashimTheArab committed:
update doc

RestartFU commented on server/block/redstone_torch.go:
```diff
@@ -0,0 +1,207 @@
+	return false
+}
+
+// NeighbourUpdateTick is called when a neighboring block is updated.
```
keepin' it bri'ish

```suggestion
// NeighbourUpdateTick is called when a neighbouring block is updated.
```

RestartFU commented on server/block/redstone_wire.go:
```diff
@@ -0,0 +1,237 @@
+		blockPower = max(blockPower, tx.RedstonePower(neighbourPos, side, false))
+
+		if side.Axis() == cube.Y {
+			// Only check horizontal neighbors from here on.
```
```suggestion
			// Only check horizontal neighbours from here on.
```

HashimTheArab committed:
Update server/block/redstone_torch.go

Co-authored-by: RestartFU <45609733+RestartFU@users.noreply.github.com>

HashimTheArab committed:
Update server/block/redstone_wire.go

Co-authored-by: RestartFU <45609733+RestartFU@users.noreply.github.com>

HashimTheArab committed:
refactor: move RedstoneUpdater interface from block to world

HashimTheArab committed:
fix: torch burnout logic only applies on bedrock when torch burns itself out, also simplify code and docs

HashimTheArab committed:
refactor: simplify note blocks and add docs

HashimTheArab committed:
perf: significantly optimize redstone.go

HashimTheArab committed:
refactor(redstone.go): simplify connection logic

HashimTheArab committed:
refactor(redstone torch): simplify logic

HashimTheArab committed:
doc

HashimTheArab committed:
fix: better design with explicit WeakBlockPowerer

HashimTheArab committed:
fix: extra updateDirectionalRedstone call in lever

HashimTheArab committed:
fix: add RedstonePowerRelayer interface instead of relying on transparent. Glowstone was incorrectly marked as transparent.
TODO: many blocks in dragonfly are incorrectly marked as transparent. In a follow-up PR fix all of those and have them implement RedstonePowerRelayer if necessary

HashimTheArab committed:
Preserve redstone propagation when block traits diverge

  Dust neighbour updates can change effective output direction without
  changing numeric power, so unchanged-power wire updates now still walk
  the wire network after passing the redstone update handler. Glowstone
  also no longer uses light transparency to opt out of redstone relay;
  it keeps normal skylight blocking and declares the redstone relay rule
  directly.

HashimTheArab committed:
fix: remove redundant lookup

HashimTheArab committed:
fix: blockPower and wirePower variable names are inverted

NopeNotDark reviewed (approved):
Reviewed as working (05/03/2026 - 5:31:48 PM EST)

didntpot committed:
Merge branch 'master' into feature/basic-redstone

TrixNEW reviewed (approved):
Reviewed as working (05/03/2026 - 5:46:23 PM EST)

HashimTheArab committed:
Resolve redstone PR against registry refactor

Keep the basic redstone interfaces from PR #1095 while accepting upstream's instance-scoped block registry split, so the feature branch can merge cleanly with current master.

Constraint: PR #1095 must merge into current df-mc/dragonfly master and be pushed to xNatsuri/dragonfly:feature/basic-redstone
Rejected: Pushing the one-parent local 'merge master' commit | it would make upstream changes look like ordinary PR changes
Confidence: high
Scope-risk: moderate
Directive: Keep redstone interfaces in block.go, but keep registry logic in BlockRegistry/DefaultBlockRegistry
Tested: go test ./...
Not-tested: in-game redstone behavior smoke test

HashimTheArab committed:
Expose redstone state under world package

Move transient redstone runtime state out of server/internal because world.Tx exposes it through Redstone(), making the importable world/redstone path the clearer public boundary.

Constraint: Tx.Redstone() returns the state type, so the package path should not be internal-only
Rejected: Keeping server/internal/redstone | public world API would expose an internal type path
Confidence: high
Scope-risk: narrow
Directive: Keep redstone runtime bookkeeping under server/world/redstone while world.Tx remains the access point
Tested: go test ./...
Not-tested: in-game redstone behavior smoke test

HashimTheArab committed:
Keep torch burnout recovery tied to real block updates

Burned-out redstone torches should not relight from calculation-only redstone propagation around disconnected dust. Move burnout recovery to the neighbour-update path and let redstone propagation leave burned-out torches inert until an actual adjacent block update occurs.

Constraint: Existing redstone wire propagation intentionally walks a broad Paper/Eigencraft-style neighbourhood.

Rejected: Rewriting wire propagation to skip broad non-wire updates | too large for the reported torch burnout recovery bug and risks changing unrelated redstone update order.

Confidence: high

Scope-risk: narrow

Directive: Do not reintroduce burnout recovery in RedstoneUpdate unless redstone propagation can distinguish real neighbour block updates from calculation-only visits.

Tested: go test ./server/block -run 'TestDisconnectedWireUpdateDoesNotRecoverBurnedOutTorch|TestNeighbourBlockUpdateRecoversBurnedOutTorch' -count=1; go test ./server/block/... -count=1; go test ./...; git diff --check

Not-tested: In-game Bedrock client redstone contraption smoke test

HashimTheArab committed:
Merge branch 'master' into feature/basic-redstone

HashimTheArab committed:
Recover burned-out torches from input-side updates

Torch burnout recovery needs to distinguish an unrelated wire-walk visit from a redstone update that actually came from the torch's input block. Carry the update source through redstone propagation and allow burned-out torches to recover only when that source is the attached block or directly beside it.

Constraint: Redstone wire propagation still performs a broad Paper/Eigencraft-style walk that can visit disconnected receivers.

Rejected: Restoring unconditional RedstoneUpdate burnout recovery | reopens disconnected dust waking unrelated burned-out torches.

Confidence: medium

Scope-risk: moderate

Directive: Preserve update source propagation for torch burnout logic unless the redstone update API grows an explicit source parameter.

Tested: temporary local block tests for disconnected dust and input dust recovery; go test ./server/block/... -count=1; go test ./...; git diff --check

HashimTheArab committed:
bug fixes

HashimTheArab committed:
Prevent stale redstone state from mutating replaced blocks

  Constraint: Scheduled redstone torch ticks may execute after the original torch
  block has been broken or replaced.
  Rejected: Trusting scheduled block receiver state | it can resurrect stale torch
  blocks.
  Confidence: high
  Scope-risk: narrow
  Directive: Reload mutable scheduled block state from the live world before writing
  it back.
  Tested: go test ./...

  If this commit includes the lever/wire/burnout fixes too, use:

  Fix redstone review regressions before PR merge

  Constraint: Review found stale torch ticks, wire placement success reporting, lever
  powered drops, and torch burnout threshold issues.
  Rejected: Broad break-handler ordering changes | container inventory handlers use
  the original block value and were covered separately.
  Confidence: high
  Scope-risk: narrow
  Directive: Keep scheduled redstone mutations guarded by live block checks.
  Tested: go test ./...

HashimTheArab committed:
clarify doc

HashimTheArab committed:
Align BFS redstone wire power with direct calculation

Constraint: Redstone wire has both single-wire and BFS update paths that must agree on vertical transmission rules.
Rejected: Keeping separate BFS neighbour-index power logic | it can diverge from the direct calculation for supported transparent/step-up cases.
Confidence: high
Scope-risk: narrow
Directive: Keep future redstone wire power rule changes in the shared calculation path.
Tested: go test ./...
Not-tested: In-game redstone fixture comparison against Bedrock/Java reference worlds.
